Feat/#1855 file entity page

ciaranschutte opened the pull request:
**Description of changes**

Adds file entity page to platform ui
- behind feature flag: FEATURE_FILE_ENTITY_ENABLED
- does page level gql query check to make sure user is navigating to valid file entity

**Type of Change**

- [ ] Bug
- [ ] Styling
- [x] New Feature

**Checklist before requesting review:**

- [x] Matches design:

  - component sizes, spacing, and styles
  - font size, weight, colour
  - spelling has been double checked

- [ ] New uikit components have Storybook stories
- [ ] Feature is minimally responsive.
- [ ] Add copyrights to new files
- [x] Manual testing of UI feature.
- [ ] Selenium test is completed and passing.

ciaranschutte commented on pages/file/[fileId].tsx:
```diff
@@ -0,0 +1,70 @@
+  const { fileId } = usePageQuery<{ fileId: string }>();
+  const filters = sqonBuilder.has('object_id', fileId).build();
+
+  // small query to ensure the fileId is valid
```
check if the fileId in the url is valid otherwise show an error page

ciaranschutte commented on components/pages/file-entity/FileCardsLayout.tsx:
```diff
@@ -38,11 +38,13 @@ const FileCardsLayout: React.ComponentType<{
           <AssociatedDonors donors={fileData.donorRecords} />
         </PaddedColumn>
       </PaddedRow>
+      {/*
```
comment out complete but inactive section for now

joneubank replied:
Why is this inactive and when will it be activated? Could you leave a code comment?

ciaranschutte replied:
@joneubank might need RDPC api for this, might not. Spec not ready yet. 
Comment as vague as this may be pointless?

ciaranschutte replied:
Just removing the unused code.
It should be clear from ticket comments that we have that last bit of UI already made.

It's the same flow has this ticket is.
Taking UI already made and adding it to a page structure.

Therefore a scattering of TODOs is not needed,it's just cruft

ciaranschutte commented on components/pages/file-entity/useEntityData.tsx:
```diff
@@ -0,0 +1,92 @@
+      summary,
+      dataAnalysis,
+      donorRecords,
+      // fileRecords: dummyFileRecords,
```
comment out complete but inactive section for now

joneubank replied:
code comment to explain, maybe a TODO

ciaranschutte commented on pages/file/[fileId].tsx:
```diff
@@ -0,0 +1,70 @@
+  if (!loading && !isValidEntity) {
+    setLoaderShown(false);
+    const err = new Error('Page Not Found') as Error & { statusCode?: number };
+    err[ERROR_STATUS_KEY] = 404;
```
returning <Error> to show an error page
usually we are throwing an error but this leaves a blank page for user

ciaranschutte commented on pages/file/[fileId].tsx:
```diff
@@ -0,0 +1,70 @@
+
+  const { setLoading: setLoaderShown, isLoading: isLoaderShown } = useGlobalLoadingState();
+
+  if (!loading && !isValidEntity) {
```
@anncatton  do I have to check if a user is loggingOut too?
I saw similar logic on the submission pages

anncatton replied:
`isLoggingOut` was added to prevent the `useProgramCheckEffect` getting called and throwing an error when logout was initiated but the user was still on a page that was trying to react to the token state...has any of the logic surrounding that changed? It's been awhile since that got put in

ciaranschutte replied:
No logic has changed around that in this PR.
We've building fairly common patterns in those page component so just making sure I have everything covered.

anncatton replied:
I'm thinking, if a file entity can have controlled access, it would be good to do the `isLoggingOut` check, so you don't get an unnecessary 403, to prevent the page from trying to fetch a file after the token has been removed on logout

ciaranschutte commented on components/pages/file-entity/index.stories.tsx:
```diff
@@ -19,10 +19,64 @@
+import { FileTitleBar } from './FileTitleBar';
+import NavBar from '../../NavBar';
 
 const FileRepositoryTableStories = storiesOf(`${__dirname}`, module).add('Basic', () => {
```
Data fetching happens in the component.
There's no point changing the component to suit the storybook usage so I've pretty much copied the page in here with the dummy data.

joneubank commented on components/pages/file-entity/useEntityData.tsx:
```diff
@@ -0,0 +1,92 @@
+      return { programShortName: null, data: null, loading: false };
+    }
+
+    const entity = get(data, 'file.hits.edges[0].node');
```
This could come back as undefined, and then all of the property references below will throw errors. We need to handle that missing value case.

joneubank commented on components/pages/file-entity/useEntityData.tsx:
```diff
@@ -0,0 +1,92 @@
+import { useQuery } from '@apollo/react-hooks';
```
Copyright notice added to files you modify? Plzkthx?

anncatton commented on components/pages/file-entity/FileTitleBar.tsx:
```diff
@@ -4,13 +4,20 @@ import TitleBar from 'uikit/TitleBar';
 }> = ({ programShortName, fileId, isUserLoggedIn }) => {
   const theme = useTheme();
+  const { GATEWAY_API_ROOT } = getConfig();
+  const filter = sqonBuilder.has(FileCentricDocumentField['object_id'], fileId).build();
```
sqon builder all over the file repository now :)

ciaranschutte commented:
@joneubank addressed your feedback. 
I just removed unused code instead of sprinkling TODOs.
We are aware the component is there for future use, I think that's sufficient.

joneubank commented on components/pages/file-entity/useEntityData.tsx:
```diff
@@ -47,12 +47,12 @@ const useEntityData = ({ fileId }: { fileId: string }): EntityData => {
-    if (!data) {
+    const entity = get(data, 'file.hits.edges[0].node', null);
+
+    if (!entity) {
```
👍

joneubank reviewed (approved):
Thanks Ciaran!
